ext/page_log/palite/palite.cpp

```diff
@@ -0,0 +1,1499 @@
+            // The database is opened for reading and writing,
+            // and is created if it does not already exist.
+            SQLITE_OPEN_READWRITE |
+            SQLITE_OPEN_CREATE |
+            // Use the "multi-thread" threading mode.
+            // Cannot share this connection between threads.
+            SQLITE_OPEN_NOMUTEX;
+
+        SQL_CALL_OPEN(sqlite3_open_v2, db_path.c_str(), db, flags, nullptr);
+        LOG_DEBUG("Opened SQLite database: {}", *db);
+
+        std::call_once(db_init, [this](sqlite3* d){ init_db(d); }, *db);
+
+        SQL_CALL_CHECK(*db, sqlite3_busy_handler, *db, &db_busy_handler,
+            static_cast<void*>(this));
+
+        std::scoped_lock lock(db_mutex);
+        close_connections.push_back([this, db](){ close_db(db); });
+
+        return *db;
+    }
+
+    void init_db(sqlite3* db) {
+        static const char* create_statements[] = {
+            R"(CREATE TABLE IF NOT EXISTS pages (
+                table_id INTEGER NOT NULL,
+                page_id INTEGER NOT NULL,
+                lsn INTEGER NOT NULL,
+                backlink_lsn INTEGER NOT NULL,
+                base_lsn INTEGER NOT NULL,
+                flags INTEGER NOT NULL,
+                encryption STRING NOT NULL,
+                timestamp_materialized_us INTEGER NOT NULL,
+                page_data BLOB,
+                PRIMARY KEY (table_id, page_id, lsn)
+            );)",
+            R"(CREATE TABLE IF NOT EXISTS globals (
+                key INTEGER NOT NULL,
+                val INTEGER NOT NULL,
+                PRIMARY KEY (key)
+            );)",
+            R"(CREATE TABLE IF NOT EXISTS checkpoints (
+                lsn INTEGER NOT NULL,
+                timestamp INTEGER NOT NULL,
+                checkpoint_metadata BLOB,
+                PRIMARY KEY (lsn, timestamp)
+            );)",
+            // These keys correspond to the GlobalKey enumeration.
+            // Key 0: LSN, 1 will be used next
+            "INSERT OR IGNORE INTO globals(key, val) VALUES (0, 1);",
+            // TODO: remove if not needed
+            // Key 1: Checkpoint completed
+            "INSERT OR IGNORE INTO globals(key, val) VALUES (1, 0);",
+            // Key 2: Checkpoint started
+            "INSERT OR IGNORE INTO globals(key, val) VALUES (2, 0);"
+        };
+
+        for (const char* sql : create_statements) {
+            SQL_CALL_CHECK(db, sqlite3_exec, db, sql, nullptr, nullptr, nullptr);
+        }
+
+        LOG_DEBUG("SQLite database schema initialized");
+    }
+
+#define SQ_CHECK(func, ...) SQL_CALL_CHECK(db_instance(), func, __VA_ARGS__)
+
+    using StatementArray = std::array<sqlite3_stmt*, Statement::COUNT>;
+    using StatementPtr = std::unique_ptr<sqlite3_stmt, std::function<decltype(sqlite3_reset)>>;
+
+    StatementPtr db_statement(Statement stmt) {
+        thread_local static StatementArray statements;
+        if (statements[0] == nullptr) {
+            prepare_statements(statements);
```

Comment:
This makes me a little nervous to do the prepares lazily.  I don't know if we can guarantee that the first calls will be nicely sequential.  Maybe call `prepare_statements()` after the database is initialized and tables created?